### PR TITLE
Fix 25798 : Allow "partial/prefix" filter for health_status, exec_create and exec_start events

### LINKED ISSUE #25798: Event filtering on "exec_start", "exec_create", and "health_status" is broken
**Output of `docker version`:**

```
Client:
 Version:      1.12.1-rc1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   7889dc7
 Built:        Fri Aug 12 18:35:53 2016
 OS/Arch:      darwin/amd64
 Experimental: true

Server:
 Version:      1.12.1-rc1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   7889dc7
 Built:        Fri Aug 12 18:35:53 2016
 OS/Arch:      linux/amd64
 Experimental: true
```

**Steps to reproduce the issue:**

In one terminal:

```bash
docker events --filter event=health_status
```

In another terminal;

```bash
docker run --name=test -d --health-cmd='stat /etc/passwd || exit 1' --health-interval=2s   busybox sleep 1d
```

**Describe the results you received:**

```bash
$ docker events --filter event=health_status
<no output>

$ docker events --filter event=exec_create
<no output>


$ docker events --filter event=exec_start
<no output>
```

**Describe the results you expected:**

```bash
$ docker events --filter event=health_status

2016-08-17T12:10:34.337301145Z container health_status: healthy da1bab7bc91a424cd8e0571808c1b4638a903a8508eb6612e9cdfc49f9dde555 (image=busybox, name=test)


$ docker events --filter event=exec_create

2016-08-17T12:10:34.198347963Z container exec_create: /bin/sh -c stat /etc/passwd || exit 1 da1bab7bc91a424cd8e0571808c1b4638a903a8508eb6612e9cdfc49f9dde555 (image=busybox, name=test)
2016-08-17T12:10:36.338156681Z container exec_create: /bin/sh -c stat /etc/passwd || exit 1 da1bab7bc91a424cd8e0571808c1b4638a903a8508eb6612e9cdfc49f9dde555 (image=busybox, name=test)


$ docker events --filter event=exec_start

2016-08-17T12:10:34.198549073Z container exec_start: /bin/sh -c stat /etc/passwd || exit 1 da1bab7bc91a424cd8e0571808c1b4638a903a8508eb6612e9cdfc49f9dde555 (image=busybox, name=test)
2016-08-17T12:10:36.338214034Z container exec_start: /bin/sh -c stat /etc/passwd || exit 1 da1bab7bc91a424cd8e0571808c1b4638a903a8508eb6612e9cdfc49f9dde555 (image=busybox, name=test)
```

**Additional information you deem important (e.g. issue happens only occasionally):**

I tried to reproduce on older versions of docker, and this seems to work on Docker 1.8, but fails on Docker 1.9

I debugged this issue, and as far as I can see, the cause of this is that, unlike other events, the `event.Action` for `exec_create`, `exec_start`, and `health_status` contains additional output. This is the content of `event.Action` of events fired by the example container;

```
create
connect
start
exec_create: /bin/sh -c stat /etc/passwd || exit 1
exec_start: /bin/sh -c stat /etc/passwd || exit 1
health_status: healthy
```

The filter executes an *exact* match on `event.Action` (`ef.filter.ExactMatch("event", ev.Action)`), and fails to recognize the events due to this extra output;

See [daemon/events/filter.go#L21)](https://github.com/docker/docker/blob/297745b1cdfb79a12986465f8150e22b0018c640/daemon/events/filter.go#L21), and [vendor/src/github.com/docker/engine-api/types/filters/parse.go#L209-L219](https://github.com/docker/docker/blob/75109b32db4c23f469b57fb99300fc5d4a689dce/vendor/src/github.com/docker/engine-api/types/filters/parse.go#L209-L219)

Possibly we can change the `ef.filter.ExactMatch()` with `ef.filter.FuzzyMatch()`, however that would result in `--filter event=exec` to match both `exec_start` and `exec_create` (not sure we want that).

If we want this filter to be an "exact" match, we could filter on;

1. Exact match (for, e.g. `start`)
2. A *prefix* match for `exec_start:` (the event name provided, but with a *colon* (`:`) after

This would prevent `exec` to match `exec_start`.

### PR DESCRIPTION
**Pull request by vdemeester**
Fixes #25798, by using partial filtering for `health_status`, `exec_create` and `exec_start` events if there is no `:` in the filter 🐅.

<del>When we filter event, we do an exact match. Thus if the Action content is dynamic, we are filtering out some events. The Attributes field is there to pass additional data with events.</del>

<del>Fixing exec_create, exec_start and health_status event in that matter. A possible follow-up of this would be to force a type for Action in these methods to make sure we pass predefined actions.</del>

**\- How to verify it**

In one terminal:

``` bash
docker events --filter event=health_status
```

In another terminal;

``` bash
docker run --name=test -d --health-cmd='stat /etc/passwd || exit 1' --health-interval=2s   busybox sleep 1d
```

**\- Expected**

``` bash
$ docker events --filter event=health_status

2016-08-17T12:10:34.337301145Z container health_status: healthy da1bab7bc91a424cd8e0571808c1b4638a903a8508eb6612e9cdfc49f9dde555 (image=busybox, name=test)


$ docker events --filter event=exec_create

2016-08-17T12:10:34.198347963Z container exec_create: /bin/sh -c stat /etc/passwd || exit 1 da1bab7bc91a424cd8e0571808c1b4638a903a8508eb6612e9cdfc49f9dde555 (image=busybox, name=test)
2016-08-17T12:10:36.338156681Z container exec_create: /bin/sh -c stat /etc/passwd || exit 1 da1bab7bc91a424cd8e0571808c1b4638a903a8508eb6612e9cdfc49f9dde555 (image=busybox, name=test)


$ docker events --filter event=exec_start

2016-08-17T12:10:34.198549073Z container exec_start: /bin/sh -c stat /etc/passwd || exit 1 da1bab7bc91a424cd8e0571808c1b4638a903a8508eb6612e9cdfc49f9dde555 (image=busybox, name=test)
2016-08-17T12:10:36.338214034Z container exec_start: /bin/sh -c stat /etc/passwd || exit 1 da1bab7bc91a424cd8e0571808c1b4638a903a8508eb6612e9cdfc49f9dde555 (image=busybox, name=test)
```

**\- Todo**
- [x] Add an integrations test for that.
- [x] Decide if the map key are the right one for these cases.

/cc @thaJeztah @icecrime @tiborvass @stevvooe @calavera @cpuguy83 

🐸

Signed-off-by: Vincent Demeester vincent@sbr.pm

### CODE REVIEW

**Comment by thaJeztah** [comment]
"status" and "command" sound good to me

**Comment by jhorwit2** [comment]
Thoughts on changing `status` since it's a [deprecated message field](https://github.com/docker/engine-api/blob/master/types/events/events.go#L32)? This PR would mean we have a deprecated `status` and a non-deprecated `status` for `container` events.

**Comment by vdemeester** [comment]
@jhorwit2 good point :sweat_smile:

File: daemon/health.go [hunk 1]
```diff
@@ -136,7 +138,9 @@ func handleProbeResult(d *Daemon, c *container.Container, result *types.Healthch
 	}
 
 	if oldStatus != h.Status {
-		d.LogContainerEvent(c, "health_status: "+h.Status)
+		d.LogContainerEventWithAttributes(c, "health_status", map[string]string{
```
Comment (stevvooe) [inline daemon/health.go]: This is an API change. We'll have to update consumers in swarmkit that take health events.

Comment (vdemeester) [inline daemon/health.go]: ah true :+1:

**Comment by vdemeester** [comment]
Putting the `needs-attention` flag 👼

**Comment by thaJeztah** [comment]
Design LGTM, but ping @stevvooe PTAL for the changes required in SwarmKit

**Comment by thaJeztah** [comment]
oh, actually, do we need to do something w.r.t. https://github.com/docker/docker/pull/25800#issuecomment-240417315 ?

**Comment by thaJeztah** [comment]
ping @talex5 could you have a look as well?

**Comment by talex5** [comment]
@thaJeztah: LGTM

**Comment by thaJeztah** [comment]
This is the issue we need to think about: https://github.com/docker/docker/blob/master/docs/deprecated.md#ambiguous-event-fields-in-api

Anyone have a better name?

**Comment by thaJeztah** [comment]
We discussed this in the maintainers review, and we're ok with this PR BUT we need a different name 😞

**Comment by vdemeester** [comment]
@thaJeztah tried out "state" instead of status 👼 (the *state* of the *health_status*, make probably a little more sence than the *status* of the *health_status*)

**Comment by tianon** [comment]
Do you think anyone's currently doing something like `docker events --filter 'event=health_status: healthy'` to where we might need some compatibility code to handle that (especially since that's the only existing way to consume these events easily)?

**Comment by thaJeztah** [comment]
Possibly this should be versioned in the API?

We discussed this, and we do want this change as it's really a bug fix

**Comment by thaJeztah** [comment]
We're ok with naming this "state" btw

**Comment by thaJeztah** [comment]
Discussing this further; to be really safe we should;

- version this change in the API
- implement a "match" in the client to allow backward compatibility

**Comment by vdemeester** [comment]
Sounds good to me 😇
@tianon : we kinda do it on traefik so I guess several project does so I do agree we should have some backward compatibility thingy

File: api/server/router/system/system_routes.go [hunk 2]
```diff
@@ -143,6 +145,40 @@ func (s *systemRouter) getEvents(ctx context.Context, w http.ResponseWriter, r *
...
+}
+
+func convertEventFilter(ef filters.Args, event, name, attribute string) {
+	if strings.HasPrefix(event, name) {
+		ef.Del("event", event)
+		ef.Add("event", name)
+		values := strings.SplitN(event, ":", 2)
+		if len(values) == 2 {
+			value := strings.TrimSpace(values[1])
+			ef.Add("label", fmt.Sprintf("%s=%s", attribute, value))
```
Comment (vdemeester) [inline api/server/router/system/system_routes.go]: This is a little bit hacky as `labels` are, in the events, store at the same level as `attributes`, so we can filter on any additionnal attributes using the label filter.

**Review by vdemeester — COMMENTED** [review]
Updated the PR to add _versionned_ filters, using a little hack :stuck_out_tongue_closed_eyes:

**Comment by vdemeester** [comment]
Also one question *arise*, is the way to only filter `health_status` event that are `healthy` for example. It is possible to do so, using the hack I'm using : `--filter event="health_status: healthy"` would now be `--filter event=health_status --filter label=healthy` — but it feels really weird to be able to do so with labels (and it's undocumented too).

**Comment by vdemeester** [comment]
Let's redo that in a less *intrusive* way. If `health_status` is provided as an event filter (without a `:`), then we'll filter `health_status: …` event using *prefix* instead of full match. It does not require any changes of the API and let's us take time to refactor some part of the events 👼

**Comment by vdemeester** [comment]
Updated to not change the API, or hacking around at all.
/cc @thaJeztah @tiborvass @cpuguy83 👼

File: daemon/events/filter.go [hunk 3]
```diff
@@ -29,6 +29,24 @@ func (ef *Filter) Include(ev events.Message) bool {
 		ef.matchLabels(ev.Actor.Attributes)
 }
 
+func (ef *Filter) matchEvent(ev events.Message) bool {
+	// #25798 if an event filter contains either health_status, exec_create or exec_start without a colon
+	// Let's to a FuzzyMatch instead of an ExactMatch.
+	if ef.filterContains("event", map[string]struct{}{"health_status": {}, "exec_create": {}, "exec_start": {}}) {
+		return ef.filter.FuzzyMatch("event", ev.Action)
```
Comment (thaJeztah) [inline daemon/events/filter.go]: We should really rename `FuzzyMatch` to `PrefixMatch` as it's not doing anything "fuzzy", only optimizing a prefix match

**Review by thaJeztah — CHANGES_REQUESTED** [review]
changes LGTM, but a test would be nice to have

File: daemon/events/filter.go (hunk 3 above)
Comment (vdemeester) [inline daemon/events/filter.go]: That's true :+1:

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by cpuguy83** [comment]
LGTM

**Comment by mlilback** [comment]
vdemeester: what is meant by "It does not require any changes of the API and let's us take time to refactor some part of the events"?

I'm running into this problem when observing exec_start via the API. Was an issue created for this refactoring?

**Pull request merged**